### DISCUSSION TITLE
Weather widget TODO list (JavaScript)

### DISCUSSION THREAD

**Discussion by valinet** [valinet/ExplorerPatcher#755] [Announcements]
Current status:

✔️ **Dark mode support**
✔️ **Better support for icons on light backgrounds**
✔️ **Support some night time icons**
✔️ **Determine sunset and sunrise times**
❌ **Support Bing as provider**
❌ **Multiple locations**

I consider this ready to ship as a stable version, once the code base stabilizes a bit; functionality for v1 is implemented.

_____

Hi

As some of you may be aware, a new feature has been recently been implemented in ExplorerPatcher for the Windows 10 taskbar: [a weather widget](https://github.com/valinet/ExplorerPatcher/wiki/Weather). The implementation leverages Microsoft's WebView2 component to scrape and display the weather widget from Google Search. For interacting with the web page, client-side JavaScript is executed, performing a host of tasks, like extracting, scaling and converting the weather icon from the web page to an array of bytes and much more.

Taking this into consideration, there is now this opportunity to write code in JavaScript for ExplorerPatcher as well. Thus, I launch an invitation for interested parties in contributing towards the following objectives, with JavaScript code. These kind of have to make their way into the code anyway before the weather widget ships in a release, so eventually I ought to work on them myself, but I was thinking maybe some of you are interested in contributing and helping accelerate development on this. So, specifically:

✔️ **Dark mode support**

**Implemented as of 42.4, see https://github.com/valinet/ExplorerPatcher/discussions/755#discussioncomment-2076429**

There are 2 parts to it:

1. Supporting dark mode for the window that hosts the flyout itself. That's easy, we already have a working implementation for that in the Properties window, we just use the code from there in the C land of things.
2. Supporting dark mode in the weather widget (JavaScript land). 

This means supporting dark mode on the Google search web page. Again, there are 2 aspects when it comes to this:

* Google Search already supports dark mode. It's set via the preferences, and then that sets a cookie with an id, and that id gets dark mode enabled associated with it. There doesn't seem to be a URL parameter to force dark mode, that would have been ideal. So yeah, some work in this area, determining how to force dark mode like this. Or if someone knows a way to hack the URL to get us dark mode directly, shout out, please. Or:
* Alter the CSS on the page manually, applying dark mode manually after the data has loaded, by altering the elements' CSS. This would be it if the above fails no matter what.

So yeah, contributions of some JavaScript that does something about this after the content has loaded would be good.

✔️ **Better support for icons on light backgrounds**

**Implemented as of 42.30. Added an option to use the stock Microsoft icons which have better contrast with the light taskbar.**

Currently, the weather icons Google serves are not really optimized for light backgrounds. You can see that by switching the Windows theme to light for the taskbar. One solution I was thinking about is to run a flood fill on the image to identify the transparent area around the icon, and then draw a border around the middle area where the icon actually is. Idk how good that will look in practice, but it is worth a try, I think. Alternatively, idk, the icon could have a backplate, although it would be pretty ugly. Else, idk, maybe replace the icons altogether with others from some other place that serves more contrasty ones for light backgrounds... I generally want to avoid introducing dependencies on other services since those may come with quotas or generally might have less availability than Google does.

Again, suggestions are welcome.

✔️ **Support some night time icons**

**Implemented as of 42.30. Microsoft's icons support daytime and nighttime icons, so I mapped Google's ones to those based on the current time and the sunrise/sunset times at the chosen location.**

Weirdly enough, Google has the same set of icons for both day time and night time. That's a pretty weird omission from their side. It's especially weird when at night the sky's clear, so a big yellow circle displays on the taskbar. Anyway, for the time being, I'd start changing that - idk if replacing with icons from some other place would work. For that case, I was also thinking of cropping out some pixels from the circle until I get the shape of a moon. That just equates to turning the right pixels from yellow to transparent to get that. After we get that done for the "sun" icon, maybe apply the same to the few other icons that display a sun with some clouds around it...? This is in strong connection with the next one:

✔️ **Determine sunset and sunrise times**

**Implemented as of 42.30. The application uses https://github.com/Triggertrap/sun-js to determine the sunset and sunrise times for the particular location.**

So yeah, after/before the above, determine how we define day/night time. Hardcode it from 7pm to 7am? Google's widget doesn't seem to have a clue about that, and I don't really like hardcoding. Instead, something like this could be used: https://github.com/Triggertrap/sun-js, which implements some algorithm to tell us the sunrise and sunset times based on geographic coordinates and a date, so basically for any place on Earth on any day, which sounds pretty good. Only thing, the user inputs a string corresponding to a name in ExplorerPatcher, how to get geographic coordinates from that? Fortunately, if we look at the "weather.com" hyperlink in the footer of the widget (that's hidden in the flyout by EP, so check in Google search manually), the URL of that contains the geographic coordinates, so we can extract those from there and we're good to go. This might actually be a solution.

❌ **Support Bing as provider**

Bing features another weather widget, which is interesting as well: it has an animation showing the weather conditions, plus some monochrome icons (white) icons which I think can be easily converted back and forth to obtain dark/light variants. It could make for a rather nice alternative to the Google widget. The program already supports multiple provider implementations, one just needs to write a proper script to be run after the page is loaded in order to grab the data.

❌ **Multiple locations**

Offer the possibility to set a list of locations and design some UI (???) to cycle through them once you open the widget. This one's a bit more involved, it has to be discussed more in-depth first.

So yeah, these are the plans/ideas/tasks for the moment, if anyone's willing to pick something up, shout out here as well so I do not start work on that and duplicate the implementation. As I said, all of the above have to be implemented in JavaScript, they will probably go in this file: https://github.com/valinet/ExplorerPatcher/blob/master/ep_weather_host/ep_weather_provider_google_script.h as that contains the scripts that run when the weather widget has loaded and that scrapes all the data required off of it.

Thanks.

**Comment by bert07** [comment]


**Reply by valinet** [reply]
This is a TO DO thread for the weather widget, not a thread for other problems regarding it. I don’t see the reason to post here. New threads can be started by anyone for free. Please post a separate thread. Thanks.

**Comment by Anixx** [comment]
The Vista weather desktop widget can still be installed on Win11 and works well. Maybe you can look into it, how does it work and where does it get data from?

**Reply by valinet** [reply]
Thanks, interesting to know, but I kind of have the architecture figured out by now. I also prefer using Google’s data, it’s more reliable for the area I am in. But I will check the gadgets out as well, never knew they still work when it comes to Internet connectivity.

**Reply by krlvm** [reply]
Actually, the _original_ Windows Vista/7 weather gadget does not work since 2015.

**Reply by Anixx** [reply]
Several gadgets from this page work just well, although show different data:
https://gadgetsrevived.com/category/weather/

![изображение](https://user-images.githubusercontent.com/1271736/151715279-db574871-3cc5-4beb-9e22-42893fac0b79.png)

**Reply by Anixx** [reply]
@krlvm you are right, the original gadget does not work, but there are others that work.

**Reply by Anixx** [reply]
Maybe, Valentin can allow to chose between several independent services with their own icon sets and data? It is not necessary to support flyouts in all those services, just tray icon and temperature.

**Comment by krlvm** [comment]
I've just opened Google Search with a modern browser in dark theme, and it displayed in dark theme respectively, without changing any preferences or accepting cookies - I'm pretty sure it relies on `prefers-color-scheme` media query, which can be set only via developer tools.

With recent WebView2 versions, [it is possible to use developer tools protocol](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.774.44&preserve-view=true#calldevtoolsprotocolmethod). Methods parameters are in JSON.
I have not found the way to set the media query yet, but there's an experimental method [`Emulation.setAutoDarkModeOverride`](https://chromedevtools.github.io/devtools-protocol/tot/Emulation/#method-setAutoDarkModeOverride) (`methodName="Emulation.setAutoDarkModeOverride", methodParams="{\"enabled\":true}"`), it is experimental though, maybe it is related to "Force Web Contents Dark Theme" Chromium flag.

**Reply by valinet** [reply]
I tried with the flag, it still showed in white. I thought I could hack it that way as well. Anyway, maybe I missed something, but that should be, indeed, the way to go.

Something that could change on the fly would be preferable too, as that would allow it to respond when the user changes the theme in Windows.

**Reply by krlvm** [reply]
@valinet, I've found the things related to media query emulation, so you can use:
`methodName="Emulation.setEmulatedMedia"`
`methodParams="{\"media\": \"prefers-color-scheme: dark\"}"`

**Reply by valinet** [reply]
Thanks. And I issue this where? In the dev tools console? Can you try? It just takes recompiling the app with dev tools and context menus enabled, i.e. commenting out these lines:

https://github.com/valinet/ExplorerPatcher/blob/master/ep_weather_host/ep_weather_host.c#L247

https://github.com/valinet/ExplorerPatcher/blob/master/ep_weather_host/ep_weather_host.c#L255

Yeah, in the future, I will expose those as configurable methods, they might come in handy.

**Reply by krlvm** [reply]
I would definitely help if I had more experience with C++ :(
This is a method of [Chromium Developer Tools protocol](https://chromedevtools.github.io/devtools-protocol), which can be called from ICoreWebView2 as described [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.774.44&preserve-view=true#calldevtoolsprotocolmethod).
Pseudocode:
```c++
_this->pCoreWebView2->lpVtbl->CallDevToolsProtocolMethod("Emulation.setEmulatedMedia", "{\"media\": \"prefers-color-scheme: dark\"}", Callback<ICoreWebView2CallDevToolsProtocolMethodCompletedHandler>(
                [](HRESULT error, PCWSTR resultJson) -> HRESULT
                {
                    MessageBox(nullptr, resultJson, L"CDP Method Result", MB_OK);
                    return S_OK;
                }).Get());
```

**Reply by valinet** [reply]
Aha, ok, I will test this out in a bit. Thanks.

**Reply by valinet** [reply]
Yeah, I have got the thing to work, but doesn't change anything. Google search seems to ignore that. Even if I try standalone in Edge, set Windows and apps theme to dark, which makes Edge use a darker theme for the UI, if I visit Google search in InPrivate mode, it's still white, even though it says it uses the "Device default" appearance. So stupid...

**Reply by valinet** [reply]
`Emulation.setAutoDarkModeOverride` looks more promising, although I think the browser does the color replacing itself, but idk, I will switch the UI to dark mode in a bit and see how it looks, maybe it does a good job already and we could use that.

**Reply by krlvm** [reply]
It's strange, when I open Google in Incognito mode with dark theme enabled, it shows in dark because of the media query. Are you sure the DevTools call is executed before the page is loaded? In case it is being detected via JavaScript, try to execute this JavaScript code in WebView: `document.write(window.matchMedia('(prefers-color-scheme: dark)').matches)`

**Reply by valinet** [reply]
![image](https://user-images.githubusercontent.com/6503598/151707772-fb3d04f8-0397-4da5-9b07-37aad28252ce.png)

**Reply by valinet** [reply]
It seems the browser already sends that, I just changed the theme to light in Windows, the DevTools detected that and executing that `window.matchMedia...` now returns `false`.

**Reply by krlvm** [reply]
But what if try Force Dark Mode?

```cpp
_this->pCoreWebView2->lpVtbl->CallDevToolsProtocolMethod("Emulation.setAutoDarkModeOverride", "{\"enabled\":true}", Callback<ICoreWebView2CallDevToolsProtocolMethodCompletedHandler>(
                [](HRESULT error, PCWSTR resultJson) -> HRESULT
                {
                    MessageBox(nullptr, resultJson, L"CDP Method Result", MB_OK);
                    return S_OK;
                }).Get());
```

**Reply by valinet** [reply]
> `Emulation.setAutoDarkModeOverride` looks more promising, although I think the browser does the color replacing itself, but idk, I will switch the UI to dark mode in a bit and see how it looks, maybe it does a good job already and we could use that.

**Reply by valinet** [reply]
Looks promising:

![image](https://user-images.githubusercontent.com/6503598/151710085-bd87fa4e-8be4-45d2-b8d3-0779fccb7021.png)

**Reply by valinet** [reply]
Does the widget appear dark by default for you in the current version, if you say the `window.matchMedia('(prefers-color-scheme: dark)').matches` returns `true` already and you get dark mode automatically? In ExplorerPatcher I mean.

**Reply by valinet** [reply]
It clearly works using `Emulation.setAutoDarkModeOverride` in the output that I get, you can see how the clouds are darker gray than normal. I mean, it looks fine, maybe the selection rectangle for the day should be manually adjusted to a more lighter gray so it's better viewed. Other than that, yeah, this is kind of what can be achieved.

**Reply by krlvm** [reply]
Sorry, I don't have a Windows 11 instance at the moment. Also, after some research, I can say that Google ignores user identity and will not appear in dark theme when you use igu=1 as URL param, but otherwise it will not be able to be embedded into an iframe until you disable CORS.

**Reply by valinet** [reply]
`Emulation.setEmulatedMedia` now works when invoked via `_this->pCoreWebView2->lpVtbl->CallDevToolsProtocolMethod`. The problem was that the command was not correct; the correct command for what we want means passing this string: `{\"features\": [ { \"name\": \"prefers-color-scheme\", \"value\": \"dark\" }]}`. I can now check using the F12 dev tools that the `window.matchMedia('(prefers-color-scheme: dark)').matches` changes as I change the theme in the UI. The problem is, the page still ignores that... :(

**Reply by krlvm** [reply]
As I said before, it apparently will ignore the media query until the `igu` URL param is present

**Reply by valinet** [reply]
https://github.com/valinet/ExplorerPatcher/blob/master/ep_weather_host/ep_weather_provider_google_html.h#L42 `igu=1` isn't that okay?

**Reply by valinet** [reply]
![image](https://user-images.githubusercontent.com/6503598/151716214-9498534b-32d3-4abe-a69e-2bbcbad7dd8a.png)

**Reply by krlvm** [reply]
When URL parameter `igu` is present, Google allows to embed itself into iframe, but with limitations which prevent dark theme from being applied, https://github.com/valinet/ExplorerPatcher/discussions/755#discussioncomment-2077120

**Reply by valinet** [reply]
I know that the user is disconnected, that's why I said to try in InPrivate. When in InPrivate in Edge, no `igu=1` is set, and it still doesn't work, with the user still being logged off. The site says it uses the "device default" theme, but it just ignores that and displays white. That's what I was saying all along.

**Reply by valinet** [reply]
Gave up on using an iframe, I now display the page directly and adjust the bounds of the `WebView2` component in order to display only the weather portion of the page (the same thing which I was previously doing with the iframe). Indeed, in this mode, after accepting the cookie consent, the page respects the `prefers-color-scheme` value. The result is pretty good:

![image](https://user-images.githubusercontent.com/6503598/151724555-58acd452-e6ea-4b06-bb21-1af89f28fc66.png)

Thanks @krlvm for the input, we can check this one off the list.

**Reply by krlvm** [reply]
I'm glad I could help!

**Comment by valinet** [comment]
List of Google<->Microsoft icon mappings:

```
#!/bin/bash

#https://ssl.gstatic.com/onebox/weather/64/sunny_s_cloudy.png
wget https://assets.msn.com/bundles/v1/weather/latest/PartlyCloudyDayV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/PartlyCloudyNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/sunny.png
wget https://assets.msn.com/bundles/v1/weather/latest/SunnyDayV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/ClearNightV3.svg

#https://ssl.gstatic.com/onebox/weather/64/cloudy.png
wget https://assets.msn.com/bundles/v1/weather/latest/CloudyV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/CloudyV3.svg

#https://ssl.gstatic.com/onebox/weather/64/rain_s_cloudy.png
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg

#https://ssl.gstatic.com/onebox/weather/64/cloudy_s_rain.png
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg

#https://ssl.gstatic.com/onebox/weather/64/rain_light.png
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg
wget https://assets.msn.com/bundles/v1/weather/latest/LightRainV3.svg

#https://ssl.gstatic.com/onebox/weather/64/rain.png
wget https://assets.msn.com/bundles/v1/weather/latest/ModerateRainV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/ModerateRainV2.svg

#https://ssl.gstatic.com/onebox/weather/64/rain_heavy.png
wget https://assets.msn.com/bundles/v1/weather/latest/ModerateRainV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/ModerateRainV2.svg

#https://ssl.gstatic.com/onebox/weather/64/thunderstorms.png
wget https://assets.msn.com/bundles/v1/weather/latest/ThunderstormsV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/ThunderstormsV2.svg

#https://ssl.gstatic.com/onebox/weather/64/cloudy_s_sunny.png
wget https://assets.msn.com/bundles/v1/weather/latest/MostlyCloudyDayV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/MostlyCloudyNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/partly_cloudy.png
wget https://assets.msn.com/bundles/v1/weather/latest/MostlySunnyDay.svg
wget https://assets.msn.com/bundles/v1/weather/latest/MostlyClearNight.svg

#https://ssl.gstatic.com/onebox/weather/64/sunny_s_rain.png
wget https://assets.msn.com/bundles/v1/weather/latest/RainShowersDayV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/RainShowersNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/rain_s_sunny.png
wget https://assets.msn.com/bundles/v1/weather/latest/RainShowersDayV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/RainShowersNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/fog.png
wget https://assets.msn.com/bundles/v1/weather/latest/FogV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/FogV2.svg

#https://ssl.gstatic.com/onebox/weather/64/snow.png
wget https://assets.msn.com/bundles/v1/weather/latest/LightSnowV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/LightSnowV2.svg

#https://ssl.gstatic.com/onebox/weather/64/snow_heavy.png
wget https://assets.msn.com/bundles/v1/weather/latest/HeavySnowV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/HeavySnowV2.svg

#https://ssl.gstatic.com/onebox/weather/64/snow_light.png
wget https://assets.msn.com/bundles/v1/weather/latest/LightSnowV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/LightSnowV2.svg

#https://ssl.gstatic.com/onebox/weather/64/snow_s_cloudy.png
wget https://assets.msn.com/bundles/v1/weather/latest/SnowShowersDayV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/SnowShowersNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/snow_s_rain.png
wget https://assets.msn.com/bundles/v1/weather/latest/RainSnowShowersNightV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/RainSnowShowersNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/cloudy_s_snow.png
wget https://assets.msn.com/bundles/v1/weather/latest/SnowShowersDayV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/SnowShowersNightV2.svg

#https://ssl.gstatic.com/onebox/weather/64/rain_s_snow.png
wget https://assets.msn.com/bundles/v1/weather/latest/RainSnowV2.svg
wget https://assets.msn.com/bundles/v1/weather/latest/RainSnowV2.svg
```

```
for f in *.svg; do inkscape -w 64 -h 64 "$f" -e "$f".64.png; done
for f in *.svg; do inkscape -w 128 -h 128 "$f" -e "$f".128.png; done
```

List of addresses for Microsoft icons:

**64px**

```
![RainShowersDayV2 svg](https://user-images.githubusercontent.com/6503598/156908214-0b245044-6d75-488b-8469-4671fb75d351.png)
![RainShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156908215-29f5d60e-bf47-466b-9b60-f19a78f3038b.png)
![RainSnowV2 svg](https://user-images.githubusercontent.com/6503598/156908216-b68edf4b-1355-4889-b3ac-216cf2210457.png)
![SnowShowersDayV2 svg](https://user-images.githubusercontent.com/6503598/156908217-38f9ea12-c5b8-4f35-9003-b01e9aa3df68.png)
![SnowShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156908218-81799162-052c-4624-a22d-28f772144ff8.png)
![SunnyDayV3 svg](https://user-images.githubusercontent.com/6503598/156908219-cf8c0c60-c37d-44b4-b266-c7ee5e8bd551.png)
![ThunderstormsV2 svg](https://user-images.githubusercontent.com/6503598/156908220-1561869e-1bfa-4dec-97e9-006da24672c1.png)
![ClearNightV3 svg](https://user-images.githubusercontent.com/6503598/156908221-ad8c1d1d-0e26-4831-b7a7-3903d9fcc913.png)
![CloudyV3 svg](https://user-images.githubusercontent.com/6503598/156908223-4a29a001-b0dc-487f-af2f-4da5c9fc5bc6.png)
![FogV2 svg](https://user-images.githubusercontent.com/6503598/156908225-9099c899-8d43-492c-a672-b99a666b1f1b.png)
![HeavySnowV2 svg](https://user-images.githubusercontent.com/6503598/156908226-186d9998-2e98-48fb-8ca9-5393831bbec9.png)
![LightRainV3 svg](https://user-images.githubusercontent.com/6503598/156908227-3d0b307c-8fff-42cb-8f40-b3ab3cf40d1d.png)
![LightSnowV2 svg](https://user-images.githubusercontent.com/6503598/156908228-a02fc908-0f1a-4978-af3e-47ecaaacc5a5.png)
![ModerateRainV2 svg](https://user-images.githubusercontent.com/6503598/156908229-509c4431-1e12-4e55-a00d-586a77a36eb4.png)
![MostlyClearNight svg](https://user-images.githubusercontent.com/6503598/156908230-7d9ffa4c-468b-458c-ae4f-d3b26a7f7a02.png)
![MostlyCloudyDayV2 svg](https://user-images.githubusercontent.com/6503598/156908231-0f708bb5-a757-47ff-86f2-ff90fac00031.png)
![MostlyCloudyNightV2 svg](https://user-images.githubusercontent.com/6503598/156908232-a3595fef-b6b2-411a-b0a6-b837a0932753.png)
![MostlySunnyDay svg](https://user-images.githubusercontent.com/6503598/156908234-413c1bf8-969b-4add-8c28-b1db261b0d57.png)
![PartlyCloudyDayV3 svg](https://user-images.githubusercontent.com/6503598/156908235-1f97c25b-3561-485b-b7fb-bf926ef1208c.png)
![PartlyCloudyNightV2 svg](https://user-images.githubusercontent.com/6503598/156908236-6e86fcb1-3fd4-41ff-bee6-006f01057c16.png)
![RainSnowShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156951616-941ff7d2-b93b-4650-8601-3e3e66002bc2.png)
```

**128px**

```
![PartlyCloudyDayV3 svg](https://user-images.githubusercontent.com/6503598/156949438-4e0c0e0d-67bc-4c76-b75e-e0ffcead3f48.png)
![PartlyCloudyNightV2 svg](https://user-images.githubusercontent.com/6503598/156949442-63f14d44-ec0e-40b2-aa1b-8e4a27ec10f5.png)
![RainShowersDayV2 svg](https://user-images.githubusercontent.com/6503598/156949443-062a0fa9-88c1-4e07-b6b1-8e52ff64f4f3.png)
![RainShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156949444-d3aea936-4c22-4f17-a201-02155396684d.png)
![RainSnowV2 svg](https://user-images.githubusercontent.com/6503598/156949445-60d12efa-a21d-40e0-b9a8-1b7a84e58944.png)
![SnowShowersDayV2 svg](https://user-images.githubusercontent.com/6503598/156949447-a6658710-567e-4977-9316-a80007df3076.png)
![SnowShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156949448-cd1b69af-4028-4153-8e40-288526577b58.png)
![SunnyDayV3 svg](https://user-images.githubusercontent.com/6503598/156949449-9320c6f5-15ef-4c17-9e72-740708f4828c.png)
![ThunderstormsV2 svg](https://user-images.githubusercontent.com/6503598/156949450-7e03a3f5-580e-4414-aaeb-3a0898afd1da.png)
![ClearNightV3 svg](https://user-images.githubusercontent.com/6503598/156949451-269d02a3-08cb-4237-9789-f1e60fdc723d.png)
![CloudyV3 svg](https://user-images.githubusercontent.com/6503598/156949452-f347fe27-5005-48f2-9c9a-899bb7b8825e.png)
![FogV2 svg](https://user-images.githubusercontent.com/6503598/156949454-81d5d47d-1f33-4859-a112-5a64ceb549a1.png)
![HeavySnowV2 svg](https://user-images.githubusercontent.com/6503598/156949456-04a4bdbd-ff3b-4484-bb30-8909baff8aa8.png)
![LightRainV3 svg](https://user-images.githubusercontent.com/6503598/156949458-dc66775d-8bb9-4d04-838e-7f550d305c26.png)
![LightSnowV2 svg](https://user-images.githubusercontent.com/6503598/156949459-dfe70eba-6c2c-4b1c-b51b-27c13ce7c08c.png)
![ModerateRainV2 svg](https://user-images.githubusercontent.com/6503598/156949460-7c132d89-efb7-457f-8810-9bf235f5737f.png)
![MostlyClearNight svg](https://user-images.githubusercontent.com/6503598/156949461-1f058cf3-6fdd-4aeb-80b7-68fa27b02845.png)
![MostlyCloudyDayV2 svg](https://user-images.githubusercontent.com/6503598/156949462-f50c21dd-85dd-4d9c-a4eb-516e6cddfb1f.png)
![MostlyCloudyNightV2 svg](https://user-images.githubusercontent.com/6503598/156949463-a427edfb-3d7f-4167-bd6f-f5019c482ea1.png)
![MostlySunnyDay svg](https://user-images.githubusercontent.com/6503598/156949465-54dd31c6-7e3a-464a-8e64-8b54b6fb7a65.png)
![RainSnowShowersNightV2 svg](https://user-images.githubusercontent.com/6503598/156950233-ccaadb4a-2e9a-4934-b41c-acd36a7f0d9c.png)
```


The icons can be obtained by inspecting the element of this feature from Microsoft Edge:

![image](https://user-images.githubusercontent.com/6503598/156913429-7bdbe0ae-46c9-4678-ba1f-ac51d332ee1a.png)

**Reply by valinet** [reply]
Mappings for new images (from `http://ssl.gstatic.com/onebox/weather/48/*.png` and `http://ssl.gstatic.com/onebox/weather/64/*.png`). These are the last 48 characters of the base64 of each image.

```
48 cloudy.png
gAAIgAAIgAAIgAC86wECCuvGtH3EIQAAAABJRU5ErkJggg==
48 cloudy_s_rain.png
A8DsIAvZ0J98BEAA//oBqG5a70gUjuIAAAAASUVORK5CYII=
48 cloudy_s_snow.png
EuzIGPqHhgAIgAAIAE1vnG7TKhwgdU4AAAAASUVORK5CYII=
48 cloudy_s_sunny.png
VwAEQAAEQAAEQAD+KcA3AodABBLnBp0AAAAASUVORK5CYII=
48 fog.png
EC30ZRbiv5QMwAAM8PF6AtMRIwMJwimNAAAAAElFTkSuQmCC
48 partly_cloudy.png
UD8fqAJVoApUgSrwqfwCJ6xaZshM+xMAAAAASUVORK5CYII=
48 rain_heavy.png
U4t1idb1T/0FcAF86g/VgX+0644VVAAAAABJRU5ErkJggg==
48 rain_light.png
LTAzVDExOjEzOjA0KzAyOjAw/lWSQAAAAABJRU5ErkJggg==
48 rain.png
MeRa9W/WClABKkAFiKofRnoGaQBkK9wAAAAASUVORK5CYII=
48 rain_s_cloudy.png
lFRe7S+mBtAAGkADeFsvKpKWeAy6FowAAAAASUVORK5CYII=
48 rain_s_snow.png
i/HUv5gqQAWoABUgWj/U6dmtXHN0sAAAAABJRU5ErkJggg==
48 rain_s_sunny.png
S30fyJ+YskAWyAJZIIUf+u496Ifu2YkAAAAASUVORK5CYII=
48 snow_heavy.png
cUH9DYB6xvNfDZ4GIuMbUY1SYGBaleEAAAAASUVORK5CYII=
48 snow_light.png
88Tmfn8L1e/EFVABFZA0LkcjEH9Eoxa+AAAAAElFTkSuQmCC
48 snow.png
kKRD6Xzx+qvBCyCw/ABgjEYCwIT8wQAAAABJRU5ErkJggg==
48 snow_s_cloudy.png
lfcC1L+YKkAFqAAVIFa/mfArO5CvUjAAAAAASUVORK5CYII=
48 snow_s_rain.png
1/kcsNXU/8gqQAWoAGXpGzmRTGmgvhAvAAAAAElFTkSuQmCC
48 sunny.png
NzjYbzHpLqUCFKCAJfkAq7RimK7qUtAAAAAASUVORK5CYII=
48 sunny_s_cloudy.png
MYe78v+ACIiACIiACJTEEyDCTi8sMWUSAAAAAElFTkSuQmCC
48 sunny_s_rain.png
EphMZLo+8U++KBAFpnwBEAqlYwd30/wAAAAASUVORK5CYII=
48 thunderstorms.png
DTTcSfRpAO//1L8B/hnAD9B4AcpTDEFdAAAAAElFTkSuQmCC

64 cloudy.png
AiAAAiAAAiAAAiAAl48fFVnRpiVnD+AAAAAASUVORK5CYII=
64 cloudy_s_rain.png
1S6F6f8CBIAAEAACcEP9AsAciK6YtipuAAAAAElFTkSuQmCC
64 cloudy_s_snow.png
ASAABIAAEAACQABepj8dv4hCnMCVbgAAAABJRU5ErkJggg==
64 cloudy_s_sunny.png
gAAQAAJAAAgAASAATqNfCT6dqrbpoPYAAAAASUVORK5CYII=
64 fog.png
/wsIQAACEIAABCCAb6hfRzSoz4hgnE4AAAAASUVORK5CYII=
64 partly_cloudy.png
AlAACkABKAAFoABOp1+6Bd0LJ+BorgAAAABJRU5ErkJggg==
64 rain_heavy.png
oP1rrAFoABqABqAS/QFSaJKDZ1ZY3QAAAABJRU5ErkJggg==
64 rain_light.png
LTAzVDExOjEzOjA0KzAyOjAw/lWSQAAAAABJRU5ErkJggg==
64 rain.png
FEABFEABFEABFOB3fgDsHp230RVQOwAAAABJRU5ErkJggg==
64 rain_s_cloudy.png
CQABIAAEgAAQAAJAAB6iPy4FckaI/hnJAAAAAElFTkSuQmCC
64 rain_s_snow.png
ABAAAkAACAABIAAEwIfoF/D2gpORVfhJAAAAAElFTkSuQmCC
64 rain_s_sunny.png
gAgQASJABIgAESACXpxfGkmKjZKKoRQAAAAASUVORK5CYII=
64 snow_heavy.png
+jJWB6R/m0sAEoAEIBb7BcA+hV/gbM+pAAAAAElFTkSuQmCC
64 snow_light.png
AP5lhgAEIAABCEAAAnzlBYgKVujL4rwkAAAAAElFTkSuQmCC
64 snow.png
dMq/xgqAAqAAKABW4g+Q1luByt+ugAAAAABJRU5ErkJggg==
64 snow_s_cloudy.png
SwMAAAAAAAAAAAAAAFxEL1Fe7CEOIBaKAAAAAElFTkSuQmCC
64 snow_s_rain.png
+Lc4AAAAAAAAAAAA+AGv/oIwYDUGyQAAAABJRU5ErkJggg==
64 sunny.png
Ls8fJgAAAAAAAADg1nkDlR7XfJiH1ggAAAAASUVORK5CYII=
64 sunny_s_cloudy.png
AAAAAAAAAAAAAAAA+Pf6Bm5v6/1bJGiwAAAAAElFTkSuQmCC
64 sunny_s_rain.png
+/JUWP5fQAAIAAEgAA6iPyWP6kEiMYUSAAAAAElFTkSuQmCC
64 thunderstorms.png
6n+NVQAVQAVQAfwS/QNrOJYuXyjUGAAAAABJRU5ErkJggg==
```
